File: bindings/csharp/Umicp.Tests/EnvelopeTests.cs
using Xunit;
using Umicp.Core;
using Umicp.Core.Types;
using Umicp.Core.Exceptions;

namespace Umicp.Tests;

public class EnvelopeTests
{
    [Fact]
    public void Envelope_ShouldCreateWithRequiredFields()
    {
        // Arrange & Act
        var envelope = new Envelope(
            fromId: "client-001",
            toId: "server-001",
            operation: OperationType.Data
        );

        // Assert
        Assert.Equal("client-001", envelope.FromId);
        Assert.Equal("server-001", envelope.ToId);
        Assert.Equal(OperationType.Data, envelope.Operation);
        Assert.NotNull(envelope.MessageId);
        Assert.NotEqual(default, envelope.Timestamp);
    }

    [Fact]
    public void Envelope_ShouldThrowValidationException_WhenFromIdIsEmpty()
    {
        // Arrange, Act & Assert
        Assert.Throws<ValidationException>(() =>
            new Envelope(
                fromId: "",
                toId: "server-001",
                operation: OperationType.Data
            )
        );
    }

    [Fact]
    public void Envelope_ShouldSerializeToJson()
    {
        // Arrange
        var envelope = new Envelope(
            fromId: "client-001",
            toId: "server-001",
            operation: OperationType.Data,
            messageId: "msg-123"
        );

        // Act
        var json = envelope.ToJson();

        // Assert
        Assert.Contains("\"from\":\"client-001\"", json);
        Assert.Contains("\"to\":\"server-001\"", json);
        Assert.Contains("\"operation\":\"data\"", json);
        Assert.Contains("\"message_id\":\"msg-123\"", json);
    }

    [Fact]
    public void Envelope_ShouldDeserializeFromJson()
    {
        // Arrange
        var json = @"{
            ""from"": ""client-001"",
            ""to"": ""server-001"",
            ""operation"": ""data"",
            ""message_id"": ""msg-123"",
            ""timestamp"": ""2024-01-01T00:00:00.000Z""
        }";

        // Act
        var envelope = Envelope.FromJson(json);

        // Assert
        Assert.Equal("client-001", envelope.FromId);
        Assert.Equal("server-001", envelope.ToId);
        Assert.Equal(OperationType.Data, envelope.Operation);
        Assert.Equal("msg-123", envelope.MessageId);
    }

    [Fact]
    public void Envelope_ShouldComputeHash()
    {
        // Arrange
        var envelope = new Envelope(
            fromId: "client-001",
            toId: "server-001",
            operation: OperationType.Data,
            messageId: "msg-123"
        );

        // Act
        var hash = envelope.ComputeHash();

        // Assert
        Assert.NotNull(hash);
        Assert.NotEmpty(hash);
        Assert.Equal(64, hash.Length); // SHA-256 hex string
    }

    [Fact]
    public void EnvelopeBuilder_ShouldBuildValidEnvelope()
    {
        // Arrange & Act
        var envelope = new EnvelopeBuilder()
            .FromId("client-001")
            .ToId("server-001")
            .Operation(OperationType.Request)
            .Capability("version", "1.0.0")
            .Build();

        // Assert
        Assert.Equal("client-001", envelope.FromId);
        Assert.Equal("server-001", envelope.ToId);
        Assert.Equal(OperationType.Request, envelope.Operation);
        Assert.NotNull(envelope.Capabilities);
        Assert.Equal("1.0.0", envelope.Capabilities["version"]);
        Assert.NotNull(envelope.Hash);
    }

    [Fact]
    public void EnvelopeBuilder_ShouldThrowValidationException_WhenRequiredFieldsMissing()
    {
        // Arrange
        var builder = new EnvelopeBuilder()
            .FromId("client-001");

        // Act & Assert
        Assert.Throws<ValidationException>(() => builder.Build());
    }
}

File: bindings/swift/Tests/UMICPTests/Core/EnvelopeTests.swift
import XCTest
@testable import UMICP

final class EnvelopeTests: XCTestCase {
    func testBasicEnvelopeCreation() throws {
        let envelope = try EnvelopeBuilder()
            .from("client-001")
            .to("server-001")
            .operation(.data)
            .messageId("msg-001")
            .build()
        
        XCTAssertEqual(envelope.from, "client-001")
        XCTAssertEqual(envelope.to, "server-001")
        XCTAssertEqual(envelope.operation, .data)
        XCTAssertEqual(envelope.messageId, "msg-001")
    }
    
    func testEnvelopeWithCapabilities() throws {
        let envelope = try EnvelopeBuilder()
            .from("client-001")
            .to("server-001")
            .operation(.data)
            .messageId("msg-001")
            .capabilityString("type", value: "request")
            .capabilityInt("timeout", value: 5000)
            .capabilityBool("streaming", value: true)
            .build()
        
        XCTAssertEqual(envelope.capabilities.count, 3)
        XCTAssertEqual(envelope.getCapability("type") as? String, "request")
        XCTAssertEqual(envelope.getCapability("timeout") as? Int, 5000)
        XCTAssertEqual(envelope.getCapability("streaming") as? Bool, true)
    }
    
    func testEnvelopeValidation() {
        XCTAssertThrowsError(try EnvelopeBuilder()
            .from("")
            .to("server")
            .operation(.data)
            .messageId("msg")
            .build()
        )
        
        XCTAssertThrowsError(try EnvelopeBuilder()
            .from("client")
            .to("")
            .operation(.data)
            .messageId("msg")
            .build()
        )
        
        XCTAssertThrowsError(try EnvelopeBuilder()
            .from("client")
            .to("server")
            .operation(.data)
            .messageId("")
            .build()
        )
    }
    
    func testEnvelopeSerialization() throws {
        let envelope = try EnvelopeBuilder()
            .from("client-001")
            .to("server-001")
            .operation(.data)
            .messageId("msg-001")
            .capabilityString("key", value: "value")
            .build()
        
        let json = try envelope.serialize()
        XCTAssertFalse(json.isEmpty)
        XCTAssertTrue(json.contains("client-001"))
        XCTAssertTrue(json.contains("server-001"))
    }
    
    func testEnvelopeDeserialization() throws {
        let json = """
        {
            "from": "client-001",
            "to": "server-001",
            "operation": 0,
            "message_id": "msg-001",
            "capabilities": {"key": "value"}
        }
        """
        
        let envelope = try Envelope.deserialize(from: json)
        XCTAssertEqual(envelope.from, "client-001")
        XCTAssertEqual(envelope.to, "server-001")
        XCTAssertEqual(envelope.operation, .data)
        XCTAssertEqual(envelope.messageId, "msg-001")
    }
    
    func testEnvelopeHashCalculation() throws {
        var envelope = try EnvelopeBuilder()
            .from("client-001")
            .to("server-001")
            .operation(.data)
            .messageId("msg-001")
            .build()
        
        try envelope.calculateHash()
        XCTAssertNotNil(envelope.hash)
        XCTAssertEqual(envelope.hash?.count, 64) // SHA-256 hex string
    }
    
    func testEnvelopeHashVerification() throws {
        let envelope = try EnvelopeBuilder()
            .from("client-001")
            .to("server-001")
            .operation(.data)
            .messageId("msg-001")
            .buildWithHash()
        
        XCTAssertTrue(try envelope.verifyHash())
    }
    
    func testEnvelopeWithPayload() throws {
        let payload = "Hello UMICP!".data(using: .utf8)!
        let envelope = try EnvelopeBuilder()
            .from("client")
            .to("server")
            .operation(.data)
            .messageId("msg")
            .payload(payload)
            .build()
        
        XCTAssertEqual(envelope.payload, payload)
    }
    
    func testAllOperationTypes() throws {
        for opType in OperationType.allCases {
            let envelope = try EnvelopeBuilder()
                .from("client")
                .to("server")
                .operation(opType)
                .messageId("msg-\(opType.rawValue)")
                .build()
            
            XCTAssertEqual(envelope.operation, opType)
        }
    }
    
    func testNativeTypesInCapabilities() throws {
        let envelope = try EnvelopeBuilder()
            .from("client")
            .to("server")
            .operation(.data)
            .messageId("msg")
            .capability("string", value: "test")
            .capability("int", value: 42)
            .capability("bool", value: true)
            .capability("double", value: 3.14)
            .capability("array", value: [1, 2, 3])
            .capability("dict", value: ["key": "value"])
            .build()
        
        XCTAssertEqual(envelope.capabilities.count, 6)
        XCTAssertNotNil(envelope.getCapability("string"))
        XCTAssertNotNil(envelope.getCapability("int"))
        XCTAssertNotNil(envelope.getCapability("bool"))
        XCTAssertNotNil(envelope.getCapability("double"))
        XCTAssertNotNil(envelope.getCapability("array"))
        XCTAssertNotNil(envelope.getCapability("dict"))
    }
}

